src/core/tiledmesh/qgscesiumtilesdataprovider.cpp

```diff
@@ -28,26 +28,430 @@
+{
+  QgsReadWriteLocker locker( mLock, QgsReadWriteLocker::Read );
+  auto it = mNodeMap.constFind( id );
+  if ( it != mNodeMap.constEnd() )
+  {
+    QStringList childIds;
+    const QList< QgsTiledMeshNode * > children = it.value()->children();
+    childIds.reserve( children.size() );
+    for ( QgsTiledMeshNode *child : children )
+    {
+      childIds << child->tile()->id();
+    }
+    return childIds;
+  }
+
+  return QStringList();
+}
+
+QStringList QgsCesiumTiledMeshIndex::getTiles( const QgsTiledMeshRequest &request, const QString &parentId )
+{
+  QStringList results;
+
+  std::function< void( QgsTiledMeshNode * )> traverseNode;
+  traverseNode = [&request, &traverseNode, &results]( QgsTiledMeshNode * node )
+  {
+    QgsTiledMeshTile *tile = node->tile();
+    // check filter box first -- if the node doesn't intersect, then don't include the node and don't traverse
+    // to its children
+
+    // TODO -- test intersection with request filterBox()
+
+    // TODO -- option to filter out nodes without content
+
+    if ( request.requiredGeometricError() <= 0 || tile->geometricError() <= 0 || tile->geometricError() > request.requiredGeometricError() )
+    {
+      // haven't traversed deep enough down this node, we need to explore children
+      const QList< QgsTiledMeshNode * > children = node->children();
+      for ( QgsTiledMeshNode *child : children )
+      {
+        if ( request.feedback() && request.feedback()->isCanceled() )
+          break;
+
+        traverseNode( child );
+      }
+
+      switch ( tile->refinementProcess() )
+      {
+        case Qgis::TileRefinementProcess::Additive:
+          // child add to parent content, so we must also include the parent
+          results << tile->id();
+          break;
+
+        case Qgis::TileRefinementProcess::Replacement:
+          // children replace the parent, so we skip the parent if we found children
+          if ( children.empty() )
+            results << tile->id();
+          break;
+      }
+    }
+    else
+    {
+      results << tile->id();
+    }
+
+  };
+
+  QgsReadWriteLocker locker( mLock, QgsReadWriteLocker::Read );
+  if ( parentId.isEmpty() )
+  {
+    if ( mRootNode )
+      traverseNode( mRootNode.get() );
+  }
+  else
+  {
+    auto it = mNodeMap.constFind( parentId );
+    if ( it != mNodeMap.constEnd() )
+    {
+      traverseNode( it.value() );
+    }
+  }
+
+  return results;
+}
+
+bool QgsCesiumTiledMeshIndex::tileCanRefine( const QString &id ) const
+{
+  QgsReadWriteLocker locker( mLock, QgsReadWriteLocker::Read );
+  auto it = mNodeMap.constFind( id );
+  if ( it == mNodeMap.constEnd() )
+    return false;
+
+  if ( !it.value()->children().isEmpty() )
+    return false; // already refined
+
+  const QString contentUri = it.value()->tile()->resources().value( QStringLiteral( "content" ) ).toString();
+  locker.unlock();
+
+  return contentUri.endsWith( QLatin1String( ".json" ), Qt::CaseInsensitive );
```

Comment:
The spec is a bit confusing about this.

https://github.com/CesiumGS/3d-tiles/tree/main/specification#tile-json says:

"A file extension is not required for content.uri. A content’s [tile format](https://github.com/CesiumGS/3d-tiles/tree/main/specification#core-tile-format-specifications) can be identified by the magic field in its header, or else as an external tileset if the content is JSON."

https://github.com/CesiumGS/3d-tiles/tree/main/specification#file-extensions-and-media-types also says that `.json` should be used, only to add at the end that "Explicit file extensions are optional. Valid implementations may ignore it and identify a content’s format by the magic field in its header."

So I wonder if we can just stick to `.json` test or if we should be checking magic field (which would be more complicated)

Reply:
That's rather... annoying :upside_down_face: . 

I've implemented a work around for this. If there's an extension which itself reveals definitely whether or not the content can be a tileset then we use that information. Otherwise we pessimistically report that the tile has remote children available and handle the json/non-json detection when trying to fetch the children.